teachertool/src/services/storageService.ts

```diff
@@ -192,3 +193,45 @@ export async function deleteChecklistAsync(name: string) {
         setLastActiveChecklistName("");
     }
 }
+
+// Returns undefined if it has not been set or if there was an issue.
+// Empty list means it was explicitly set to empty.
+export function getExpandedCatalogTags(): string[] | undefined {
+    try {
+        const tags = getValue(EXPANDED_CATALOG_TAGS_KEY);
+        return tags ? JSON.parse(tags) : undefined;
+    } catch (e) {
+        logError(ErrorCode.localStorageReadError, e);
+        return undefined;
+    }
+}
+
+export async function setExpandedCatalogTags(tags: string[]) {
+    try {
+        setValue(EXPANDED_CATALOG_TAGS_KEY, JSON.stringify(tags));
+    } catch (e) {
+        logError(ErrorCode.localStorageWriteError, e);
+    }
+}
+
+export async function addExandedCatalogTagAsync(tag: string) {
+    let expandedTags = getExpandedCatalogTags();
+    if (!expandedTags) {
+        expandedTags = [];
+    }
+    expandedTags.push(tag);
+    await setExpandedCatalogTags(expandedTags);
+}
+
+export async function removeExpandedCatalogTagAsync(tag: string) {
+    let expandedTags = getExpandedCatalogTags();
+    if (!expandedTags) {
+        await setExpandedCatalogTags([]);
```

Comment:
Why is it needed that we set the expanded tags to an empty list if `getExpandedCatalogTags()` returns undefined?

Reply:
If someone is trying to remove a saved expanded tag but there were none saved to begin with, it's a pretty bizarre scenario to begin with, but I decided to switch it over to an empty list (to save state as "all categories collapsed") rather than undefined, which is mostly just the initial "nothing has happened yet" state. It's pretty arbitrary handling of a shouldn't-ever-happen edge case.